packages/eslint-plugin/src/rules/no-unnecessary-condition.ts

```diff
@@ -646,22 +646,32 @@ export default createRule<Options, MessageId>({
           .getCallSignaturesOfType(
             getConstrainedTypeAtLocation(services, callback),
           )
-          .map(sig => sig.getReturnType());
+          .map(sig => sig.getReturnType())
+          .map(t => {
+            if (tsutils.isTypeParameter(t)) {
+              return checker.getBaseConstraintOfType(t);
+            }
+
+            return t;
+          });
         /* istanbul ignore if */ if (returnTypes.length === 0) {
           // Not a callable function
           return;
         }
         // Predicate is always necessary if it involves `any` or `unknown`
-        if (returnTypes.some(t => isTypeAnyType(t) || isTypeUnknownType(t))) {
+        if (
+          // https://github.com/typescript-eslint/typescript-eslint/issues/10438
+          returnTypes.some(t => !t || isTypeAnyType(t) || isTypeUnknownType(t))
+        ) {
           return;
         }
-        if (!returnTypes.some(isPossiblyFalsy)) {
+        if (!returnTypes.some(t => t && isPossiblyFalsy(t))) {
```

Comment:
Checking `t` here is unnecessary (as the `returnTypes.some` check above would return), but I preferred having this over a type assertion.

Reply:
-1 from me, since we know `t` is truthy the extra runtime `t &&` check is confusing.

I also find it kind of clunky that there are three `.some`s on the same array in rapid succession... and also a five-year-old `/* istanbul ignore if */` just before them. 

WDYT about something like this?

```ts
let reportMessage: MessageId | undefined;

for (const type of returnTypes) {
  if (!type || isTypeAnyType(type) || isTypeUnknownType(type)) {
    return;
  }

  if (isPossiblyFalsy(type)) {
    reportMessage = 'alwaysTruthyFunc';
  } else if (isPossiblyTruthy(type)) {
    reportMessage = 'alwaysFalsyFunc';
  }
}

if (reportMessage) {
  context.report({ node: callback, messageId: reportMessage });
}
```

Reply:
Yes! That's significantly easier to understand 👍 

I've made some changes since it doesn't exactly match the current implementation (a couple of failing tests). The existing code is somewhat confusing with: https://github.com/typescript-eslint/typescript-eslint/blob/ce5d3e3b5b1c2d85482c9ee6f66fed59fe04d584/packages/eslint-plugin/src/rules/no-unnecessary-condition.ts#L658-L669